### PR TITLE
Update balance calculation logic: Confirmed - unconfirmed outgoing

### PR DESCRIPTION
**Pull request by nulllpc**
# Description
`getBalance` only returns a number which is not very precise to how Bitcoin works. In order to stay as accurate as possible with only 1 number, we should return balance as confirmed balance - unconfirmed outgoing. With this formula, users can see the balance change immediately when they send fund and only see incoming balance change when it's confirmed.

- Changed the logic
- Added tests
- Updated typedefs

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1215812858094239

### CODE REVIEW

**Comment by nulllpc** [comment]
I also notice that the typedef was outdated compared to the code by a lot so I added a separate commit to address them.

**Comment by nulllpc** [comment]
Another point is that the current test setup always requires regtest and electrs to be installed which is not necessary, we should be able to test client classes without setting up those two. This should be addressed in the future

File: src/wallet-account-read-only-btc.js [hunk 1]
```diff
@@ -180,16 +178,19 @@ export default class WalletAccountReadOnlyBtc extends WalletAccountReadOnly {
...
   async getBalance () {
     await this._ensureConnected()
 
     const address = await this.getAddress()
 
-    const { confirmed, unconfirmed } = await this._client.getBalance(address)
+    const { confirmed, unconfirmedOutgoing } = await this._client.getBalance(address)
 
-    return BigInt(confirmed + (unconfirmed || 0))
+    return BigInt(confirmed - (unconfirmedOutgoing || 0))
```
Comment (jonathunne) [inline src/wallet-account-read-only-btc.js]: Edge case where the result goes negative: `unconfirmedOutgoing` counts spends of unconfirmed UTXOs too, i.e. funds that were never in confirmed in the first place.

To reproduce: fresh address receives 50,000 sats (unconfirmed), then immediately spends 20,000 of it with 29,000 change and a 1,000 fee. Blockbook reports balance: "0" plus the two pending txs, so the wallet shows `-21,000` sats.

File: src/wallet-account-read-only-btc.js [hunk 2]
```diff
@@ -180,16 +178,19 @@ export default class WalletAccountReadOnlyBtc extends WalletAccountReadOnly {
...
+   *
    * @returns {Promise<bigint>} The bitcoin balance (in satoshis).
    */
   async getBalance () {
     await this._ensureConnected()
 
     const address = await this.getAddress()
 
-    const { confirmed, unconfirmed } = await this._client.getBalance(address)
+    const { confirmed, unconfirmedOutgoing } = await this._client.getBalance(address)
```
Comment (jonathunne) [inline src/wallet-account-read-only-btc.js]: Only blockbook produces `unconfirmedOutgoing`. Electrum transports passes through the raw `blockchain.scripthash.get_balance` result, which is just `{ confirmed, unconfirmed }`

**Review by jonathunne — CHANGES_REQUESTED** [review]

File: src/wallet-account-read-only-btc.js (hunk 1 above)
Comment (nulllpc) [reply src/wallet-account-read-only-btc.js]: Corrected and added new tests to guard this

File: src/wallet-account-read-only-btc.js (hunk 2 above)
Comment (nulllpc) [reply src/wallet-account-read-only-btc.js]: I update so that we only use unconfirmedOutgoing if it's present, this would be backward compatible but a little inconsistent. I think it's good enough given that we support many types of client, and not every client gives us the same data. I also made it clear in the comment

File: src/transports/blockbook-client.js [hunk 3]
```diff
@@ -20,6 +20,41 @@
...
+
+  return pending.reduce((total, tx) => {
+    const vin = (tx.vin || []).filter(entry => !pendingTxids.has(entry.txid))
+    const spent = sumOwnedValues(vin, address)
+    if (spent === 0) return total
+
+    const change = sumOwnedValues(tx.vout, address)
+    return total + (spent - change)
+  }, 0)
+}
```
Comment (jonathunne) [inline src/transports/blockbook-client.js:94]: The chaining fix kills the negative balance, but I think this filter got a bit too greedy. It skips any vin whose parent is pending, but still credits that parent's change. That's fine when the parent is someone else's deposit, but when it's your own send the change was already counted, so now we're double-crediting it. Fix:

```suggestion
function getUnconfirmedOutgoing (transactions, address) {
  const pending = (transactions || []).filter(tx => tx.blockHeight === -1)

  const pendingDeposits = new Set(
    pending
      .filter(tx => sumOwnedValues(tx.vin, address) === 0)
      .map(tx => tx.txid)
  )

  return pending.reduce((total, tx) => {
    const vin = (tx.vin || []).filter(entry => !pendingDeposits.has(entry.txid))
    const spent = sumOwnedValues(vin, address)
    if (spent === 0) return total

    const change = sumOwnedValues(tx.vout, address)
    return total + Math.max(spent - change, 0)
  }, 0)
}```

**Review by jonathunne — CHANGES_REQUESTED** [review]

File: src/transports/blockbook-client.js (hunk 3 above)
Comment (nulllpc) [reply src/transports/blockbook-client.js:94]: I did a whole rewriting of this from scratch. The proposed solution above is good but it's still leaky: chaining can happen more than once, maybe dozens of times, and the above impl doesn't account for that.

**Comment by nulllpc** [comment]
I also added detailed descriptions in tests so that we know which cases do we need to deal with. This is a tricky issue and so we should be very careful if we implement the same logic for other clients

File: src/transports/blockbook-client.js [hunk 4]
```diff
@@ -20,6 +20,80 @@
...
+    // already accounted for by whichever tx produced it (or was never real,
+    // if that tx isn't rooted).
+    const vin = (tx.vin || []).filter(entry => !pendingTxids.has(entry.txid))
+    const spent = sumOwnedValues(vin, address)
+
+    const vout = (tx.vout || []).filter(entry => !consumedOutpoints.has(`${tx.txid}:${entry.n}`))
+    const change = sumOwnedValues(vout, address)
+
+    return total + (spent - change)
+  }, 0)
```
Comment (jonathunne) [inline src/transports/blockbook-client.js:93]: A pending send mixing a confirmed input with an unconfirmed deposit from elsewhere makes `spent - change` go negative, so `unconfirmedOutgoing` comes out negative and inflates the balance above confirmed (`confirmed=5000 -> balance=9000`). Outgoing can't be negative -- wrap the reduce in `Math.max(0, ...)`:

```js
return Math.max(0, rootedTxs.reduce((total, tx) => {
  ...
}, 0))
```

Verified: fixes the mixed case (balance -> 5000), keeps the chain cases correct, and all 20 tests stay green. Worth a test for this shape too, since none currently covers it.

Comment (nulllpc) [reply src/transports/blockbook-client.js:93]: This is a very edge case, I feel like we're going in circles a bit with these edge cases so I did a little bit of research into how Bitcoin Core handles this. 

Core doesn't do partial credit at all - a pending transaction is either fully trusted (every input is confirmed or chained from another trusted tx) or not, and if not, none of its change counts yet, only the confirmed money it spent. All-or-nothing, not partial accounting. 

I think this is a reasonable approach that satisfies 99% of cases while completely cutting out the edge cases. It's also very similar to what I already implemented. Otherwise, I have a hunch we will just keep discovering some edge cases that we have not covered and the code just bloats

Comment (nulllpc) [reply src/transports/blockbook-client.js:93]: For reference, this is the file that I read from Bitcoin Core:

https://github.com/bitcoin/bitcoin/blob/master/src/wallet/receive.cpp#L205

Got a little help from AI as well

Comment (jonathunne) [reply src/transports/blockbook-client.js:93]: Agree with this all-or-nothing approach. As long as it's documented well in the JSDocs and README.

File: src/transports/blockbook-client.js [hunk 5]
```diff
@@ -20,6 +20,87 @@
...
+  // gets no credit for its own change until the whole thing is trusted.
+  function isTrusted (tx) {
+    if (trustedTxMap.has(tx.txid)) return trustedTxMap.get(tx.txid)
+
+    // Assume untrusted while resolving, to guard against any (invalid) cycle.
+    trustedTxMap.set(tx.txid, false)
+
+    const ownedVins = (tx.vin || []).filter(entry => entry.isAddress && entry.addresses?.includes(address))
+
+    const trusted = ownedVins.length > 0 && ownedVins.every(vin => {
```
Comment (jonathunne) [inline src/transports/blockbook-client.js]: Checked this against Bitcoin Core https://github.com/bitcoin/bitcoin/blob/master/src/wallet/receive.cpp#L205-L233 : it checks every input and rejects the transaction if any parent output isn’t owned. Here we filter to `ownedVins` first. To align, I think we should instead require `ownedVins.length === vins.length`.

File: types/src/wallet-account-read-only-btc.d.ts [hunk 6]
```diff
@@ -83,21 +83,25 @@ export default class WalletAccountReadOnlyBtc extends WalletAccountReadOnly {
...
-     * @returns {Promise<BtcMaxSpendableResult>} The maximum spendable result.
+     * @returns {Promise<BtcMaxSpendableResult>} The estimated maximum spendable result.
      */
     getMaxSpendable(opts?: {
         feeRate?: number | bigint;
     }): Promise<BtcMaxSpendableResult>;
+    /**
+     * Closes any internal connection with the server.
+     */
+    dispose(): void;
```
Comment (jonathunne) [inline types/src/wallet-account-read-only-btc.d.ts:104]: Most of this type diff looks unrelated to the balance change and introduces declaration churn. Can we limit it to `BtcBalance.unconfirmedOutgoing` and any matching docs?

File: tests/wallet-account-read-only-btc.test.js [hunk 7]
```diff
@@ -209,4 +209,31 @@ describe.each([44, 84])('WalletAccountReadOnlyBtc', (bip) => {
...
+  function stubClient (balance) {
+    return {
+      connect: async () => {},
+      getBalance: async () => balance
+    }
+  }
+
+  test('should subtract unconfirmedOutgoing when the client reports it', async () => {
+    const account = new WalletAccountReadOnlyBtc(ADDRESSES[84], { network: 'regtest' })
+    account._client = stubClient({ confirmed: 100_000, unconfirmed: 999_999, unconfirmedOutgoing: 30_000 })
```
Comment (jonathunne) [inline tests/wallet-account-read-only-btc.test.js]: Could these pass the stub client through constructor config instead of assigning `account._client`? That keeps the tests on the public API.

**Review by jonathunne — CHANGES_REQUESTED** [review]

File: types/src/wallet-account-read-only-btc.d.ts (hunk 6 above)
Comment (nulllpc) [reply types/src/wallet-account-read-only-btc.d.ts:104]: I only picked the diffs related to my changes, this must have slipped through. I'll double-check

Comment (nulllpc) [reply types/src/wallet-account-read-only-btc.d.ts:104]: I have no idea how are there so many changes to the types? I'll revert them

Comment (nulllpc) [reply types/src/wallet-account-read-only-btc.d.ts:104]: Ahhh, the changes are actually accurate. The typedef was outdated, I left a comment about this here. I guess I can cherry pick that commit to another PR, but the changes are actually pretty straightforward. What would you prefer?

https://github.com/tetherto/wdk-wallet-btc/pull/99#issuecomment-4928487732

Comment (jonathunne) [reply types/src/wallet-account-read-only-btc.d.ts:104]: Ok, let's leave them in this PR.

**Review by jonathunne — APPROVED** [review]
LGTM
